### PR TITLE
ENH: Refactor LightpathMixin and all relevant devices

### PR DESCRIPTION
**Pull request by tangkong**
## Description
Reworks LightpathMixin to have a simpler interface and subscription system.
- Relevant signals are now grouped under a `SummarySignal` (a subclass of AggregateSignal)
- The only required method is now `get_lightpath_status`, which returns a dataclass contianing information needed by lightpath.
- Reworks some existing devices to implement this interface

Will have to be coordinated with a corresponding [lightpath PR](https://github.com/pcdshub/lightpath/pull/133)

This will require a significant overhaul of the existing devices that are to be used in lightpath.  Currently: 
- 177 devices are marked as active and participating in lightpath (active=True, lightpath=True)
  - Of these, only 108 implement the existing LightpathMixin
  - Several devices should be in the lightpath but are not (xpp_lodcm, xcs_lodcm in particular)
- This interface will likely require more specific device classes for branching devices, as their input and output branches will need specificity.

## Motivation and Context
Lightpath is getting an overhaul, and this new interface will be part of it

## How Has This Been Tested?
Eventually in the test suite

## Where Has This Been Documented?
This PR

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on travis
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

File: pcdsdevices/interface.py
Comment (klauer) [inline pcdsdevices/interface.py:1659]: I am somewhat out of the loop on this, so grain of salt:

I had assumed the mixin itself would go away and we'd end up with a signal like:

`lightpath = Cpt(LightpathSignal, signals=["sig1", "sig2"], status_func=get_lightpath_status)`

Or possibly just retrieving those from the parent class/instance on `__init__`.

I'm not claiming this is in any way _better_. Before updating all of the device classes it may be useful to discuss the benefits of the current approach over other options with an eye toward future maintainability.

File: pcdsdevices/signal.py
Comment (klauer) [inline pcdsdevices/signal.py:434]: Could we just return the tuple itself? Or is there good reason not to?

**Review by klauer — COMMENTED** [review]
Looking good!

File: pcdsdevices/interface.py
Comment (tangkong) [reply pcdsdevices/interface.py:1659]: I thought the mixin was a simple way to wrap up the functionality, which boils down to adding the `SummarySignal` and requiring the `get_lightpath_status` method.  I thought it was clearer this way, though I will admit that I didn't think much about a signal that sets a method.  (Component internals scare me slightly).  

I felt the main drawback of the old mixin was that the interface had many too many bits to implement (insert, remove, etc).  Is inheriting from the mixin itself a step that users find onerous? 

I guess the other question is: is the proposed API adequate?  (Outside of the implementation details)

This is the kind of feedback I was looking for, happy to consider other implementation approaches 👍

File: pcdsdevices/signal.py
Comment (tangkong) [reply pcdsdevices/signal.py:434]: I wanted to return an ophyd-compatible data type, due to some issue I had that I can't find any notes on.... 

I suppose there's no real reason to not just return the tuple.  I didn't want people accessing this signal and hoping to get any information from it, since I wasn't grouping any metadata with it (what the values correspond to)

**Review by ZLLentz — COMMENTED** [review]
I like where this is and where it's going- left verbose thoughts. I'll save my other PR review for tomorrow.

File: pcdsdevices/interface.py
Comment (tangkong) [reply pcdsdevices/interface.py:1700]: I believe it's just by name ([src](https://github.com/pcdshub/pcdsdevices/blob/06d1fc9d4a71af7b85f6688441995715597a3203/pcdsdevices/signal.py#L369))

File: pcdsdevices/interface.py
Comment (tangkong) [reply pcdsdevices/interface.py:1770]: This was more meant to be a reasonable default, which worked for the general case.  I envision these being customized per class, if we do go with this implementation

File: pcdsdevices/interface.py
Comment (tangkong) [reply pcdsdevices/interface.py:1728]: > Is that a correct understanding?

Yep, that's exactly what I had in mind.

> splitting get_lightpath_status -> (get_lightpath_status, lightpath_calc)

I like this suggestion, it would make testing easier while also limiting the number of methods users would have to define.  

> making `lp_summary` more useful

I'm wary of bundling too much of the status calculation in the signal itself, since its purpose is more to notify consumers that the signals have changed.  While in the lightpath case, we might want to run the status calculations with every update, I could imagine a case where signals update more frequently than we want to update the gui.  (though I can't imagine the calculations should be that intensive I guess)

I might also be concerned about a single signal containing semi-structured data that needs device knowledge to interpret.  What order do the signal values appear in, do we try to keep that consistent across devices, do users start to use that in other processes, etc.

File: pcdsdevices/interface.py
Comment (tangkong) [reply pcdsdevices/interface.py:1567]: I think you're right.  I can move this to `lightpath`

File: pcdsdevices/interface.py
Comment (klauer) [reply pcdsdevices/interface.py:1659]: > I thought the mixin was a simple way to wrap up the functionality, which boils down to adding the SummarySignal and requiring the get_lightpath_status method. I thought it was clearer this way, though I will admit that I didn't think much about a signal that sets a method. (Component internals scare me slightly).

I think the internals of component - however scary - aren't so necessary to understand. In this case, looking at it as "Component instantiates a `SummarySignal` that has a `.parent` that leads you back to the relevant Device instance (with the important methods)" is probably good enough?

> Is inheriting from the mixin itself a step that users find onerous?

As much as I tend to (over)use it for its code duplication-reducing aspects, I think multiple inheritance should be avoided wherever possible.

> I guess the other question is: is the proposed API adequate? (Outside of the implementation details)

Given the goals, I think it's sufficient. Including the split of `get` vs `calculate` detailed in other conversations below is important, too, I think.

Comment (tangkong) [reply pcdsdevices/interface.py:1659]: (summarizing external discussions)
After some consideration, I think a mixin is appropriate here.  The key point is that we will be storing branch information (`input_branches` and `output_branches`) externally in the happi database, and requiring them as init kwargs.  In doing so, we need to store that information and use it in the calc_lightpath_state methods; this is a pattern that lends itself to mixins.

**Comment by tangkong** [comment]
I think we've reached a consensus here about the implementation approach, and I've implemented the suggestions noted during this first review round.  

I'm going to start attempting to port over existing lightpath devices to the new interface.  Perhaps more fine tuning will come as a result.

**Comment by tangkong** [comment]
I think I'm going to push for this PR to get merged sooner than later, so we can start resolving merge conflicts and prevent this PR from becoming too lengthy.  Once the corresponding PR from lightpath is merged/tagged, maybe I can get these tests passing and this PR merged.

**Comment by tangkong** [comment]
A rebase happened, apologies for the spam

**Comment by tangkong** [comment]
I've finished applying the lightpath interface to every device marked as `lightpath=True` in the current happi database.  

LightpathMixin has relaxed its requirements slightly, allowing init without branch metadata and deferring implementation checks until `get_lightpath_state()`.  This is because there are some device classes that are both lightpath active and inactive.  Instead of creating an additional mixed-in class each time, I chose to relax the mixin.  

Container-wise, I made the `LCLSLightpathItem` branches optional, since we have several containers that service both lightpath and non-lightpath devices.  This allows the database to pass validation (w.r.t. the lightpath items), with only the lightpath-active devices getting additional metadata

Basically everything is optional now, I guess. 🤷

**Comment by tangkong** [comment]
9e7c57b restores the behavior that delayed lightpath calculations if the state objects were not initialized.  This cleanses a huge amount of errors from the lightpath gui output, so if you're testing with this branch you should pull this update

**Comment by tangkong** [comment]
Some substantial changes this month as we start to wrap up this effort.
- OffsetMirror (cc7713) logic was added, along with needed modifications to the happi container (6a9c290)
  - While the logic of each mirror is similar, but each mirror varies in which components have state positioner components.  To help with this, `calc_lightpath_state` was made to rely on helper methods that can be overloaded in future classes.  
  - Range thresholds are stored in happi as simple arrays.  There is obviously room for confusion here, but I'd love to find a better way without adding huge amounts of init kwargs and happi entries
- LODCM (b798397) logic was added.  LODCM's as currently deployed only have two variants, so separate subclasses were made. 
- LightpathState has been refactored to accept multiple outputs, merging `transmission` and `output_branch` into one field: `output` (changes made in c611124, see https://github.com/pcdshub/lightpath/pull/160)

**Comment by tangkong** [comment]
I'm sorry, I rebased.  This one gets a squash for sure

**Comment by tangkong** [comment]
Tests are passing now, and I'd like to humbly request a re-review.  The main work that was holding this up was updating the relevant devices, but there were also some changes to the `LightpathMixin` itself, mainly in an effort to make sure it was as optional as possible.  

It's a lot to sift through, thanks in advance for your patience with me here

File: pcdsdevices/mirror.py
Comment (klauer) [inline pcdsdevices/mirror.py:535]: Python syntactic sugar instead, perhaps?
```suggestion
        ins_bools = [limit[0] < x < limit[1]
```

File: pcdsdevices/mirror.py
Comment (klauer) [inline pcdsdevices/mirror.py:582]: Missing `pitch` in docstring

File: pcdsdevices/mirror.py
Comment (klauer) [inline pcdsdevices/mirror.py:776]: This same loop is here a few times; you might consider a helper that does the logic and validation:

```python
def find_matching_range_indices(ranges, value) -> List[int]:
  if np.shape...
```

File: pcdsdevices/spectrometer.py
Comment (klauer) [inline pcdsdevices/spectrometer.py:348]: TODO reminder callout

File: pcdsdevices/mirror.py
Comment (klauer) [inline pcdsdevices/mirror.py:897]: TODO reminder

File: pcdsdevices/mirror.py
Comment (klauer) [inline pcdsdevices/mirror.py:1073]: TODO reminder

**Review by klauer — COMMENTED** [review]
I added some code quality/refactor suggestions which are admittedly subjective, so feel free to take them with a grain of salt as usual.

I think this is looking really good as a sweeping first pass. It seems time to give this a real try, and we can revisit as needed.

File: docs/source/upcoming_release_notes/1028-enh_lp_mixin.rst
Comment (tangkong) [reply docs/source/upcoming_release_notes/1028-enh_lp_mixin.rst:6]: 🤦

File: pcdsdevices/lodcm.py
Comment (tangkong) [reply pcdsdevices/lodcm.py:1875]: I think that's reasonable.  I can't think of a good reason not to, at least.

File: pcdsdevices/spectrometer.py
Comment (tangkong) [reply pcdsdevices/spectrometer.py:348]: I skipped these devices because they weren't the big branching ones.  I'll make a followup PR when I get that logic sorted out.

**Comment by tangkong** [comment]
@ZLLentz  I don't know if you have time to review this, but I'm hoping to get this in before any more merge conflicts get created.  I think I've addressed most of Ken's points, leaving TO-DO's for later.

**Review by ZLLentz — APPROVED** [review]
I think it looks good, I didn't fully evaluate every line because there are so many, but since you've only touched lightpath interfaces the hutch python should continue to work as-is and lightpath should be able to move forward. I suspect we'll have tweaks to make but they won't be obvious until we run with this for a bit.

**Comment by tangkong** [comment]
Ok then, thanks everyone.  Here goes nothing

**Pull request merged**
